From 4b00ab4362324d4ea993cc357afc4024e9c064d9 Mon Sep 17 00:00:00 2001
From: Ariel D'Alessandro <ariel.dalessandro@collabora.com>
Date: Wed, 4 Dec 2024 11:24:15 -0300
Subject: [PATCH 01/79] .gitlab-ci.yml: Add initial CI pipeline for Mediatek
 Genio devices

Based on https://gitlab.collabora.com/hardware-enablement/rockchip-3588/linux/

Currently supports:
* mt8390-genio-700-evk

Signed-off-by: Ariel D'Alessandro <ariel.dalessandro@collabora.com>
(cherry picked from commit 46813efd6d3fc2fb646928d11cd64dde8367260b)
---
 .gitlab-ci.yml    | 180 ++++++++++++++++++++++++++++++++++++++++++++++
 lava/testjob.yaml |  46 ++++++++++++
 2 files changed, 226 insertions(+)
 create mode 100644 .gitlab-ci.yml
 create mode 100644 lava/testjob.yaml

diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
new file mode 100644
index 000000000000..953af748c48e
--- /dev/null
+++ b/.gitlab-ci.yml
@@ -0,0 +1,180 @@
+default:
+  image: debian:testing
+  tags:
+    - trixie
+
+stages:
+  - build
+  - test
+  - generate
+  - lava
+
+workflow:
+  rules:
+    - if: $CI_PIPELINE_SOURCE != "merge_request_event"
+
+check-devicetrees:
+  stage: build
+  tags:
+    - ultra-heavyweight
+  variables:
+    DEBIAN_FRONTEND: noninteractive
+    GIT_SUBMODULE_STRATEGY: normal
+    ARCH: arm64
+    DEFCONFIG: defconfig
+    CROSS_COMPILE: aarch64-linux-gnu-
+    DUT_LIST: "mt8370-genio-510-evk \
+               mt8390-genio-700-evk \
+               mt8395-genio-1200-evk \
+               mt8395-radxa-nio-12l "
+  before_script:
+    - apt update
+    - apt install -y devscripts
+                     build-essential
+                     crossbuild-essential-arm64
+                     bc
+                     bison
+                     flex
+                     swig
+                     python3-dev
+                     python3-pip
+                     python3-setuptools
+                     yamllint
+    - pip3 install --break-system-packages dtschema
+  script:
+    - make $DEFCONFIG
+    - make -j$(nproc) dt_binding_check
+    - DT_WARNINGS=false
+    - for DUT in ${DUT_LIST}; do
+         DT_WARNING_FILE="dt-warnings-${DUT}.txt";
+         make -j$(nproc) CHECK_DTBS=y mediatek/${DUT}.dtb 2> ${DT_WARNING_FILE};
+         if [[ $(cat ${DT_WARNING_FILE} | wc -l) > 0 ]] ; then
+             cat ${DT_WARNING_FILE};
+             DT_WARNINGS=true;
+         fi;
+      done
+    - if [ ${DT_WARNINGS} = true ] ; then exit 42; fi
+  allow_failure:
+    exit_codes:
+      - 42
+  artifacts:
+    when: on_failure
+    paths:
+      - dt-warnings-*.txt
+
+.build-debian-package:
+  stage: build
+  tags:
+    - ultra-heavyweight
+  cache:
+    when: on_success
+    key: $CI_COMMIT_REF_SLUG
+    paths:
+      - ccache
+  variables:
+    DEBIAN_FRONTEND: noninteractive
+    GIT_SUBMODULE_STRATEGY: normal
+    ARCH: amd64
+    DEFCONFIG: defconfig
+    CCACHE_BASEDIR: $CI_PROJECT_DIR
+    CCACHE_DIR: $CI_PROJECT_DIR/ccache
+  before_script:
+    - dpkg --add-architecture arm64
+    - apt update
+    - apt install -y devscripts
+                     build-essential
+                     crossbuild-essential-arm64
+                     bc
+                     bison
+                     ccache
+                     flex
+                     rsync
+                     kmod
+                     cpio
+                     libdw-dev
+                     libelf-dev
+                     libssl-dev
+                     libssl-dev:arm64
+    # Setup ccache
+    - export PATH="/usr/lib/ccache:$PATH"
+    - ccache -s
+  script:
+    - make $DEFCONFIG
+    - make -j$(nproc) $ADDITIONAL_BUILD_CMD bindeb-pkg
+    - mkdir artifacts && dcmd mv ../*.changes artifacts/
+  artifacts:
+    paths:
+      - artifacts
+
+build-arm64-debian-package:
+  extends: .build-debian-package
+  variables:
+    ARCH: arm64
+    CROSS_COMPILE: aarch64-linux-gnu-
+    ADDITIONAL_BUILD_CMD: KBUILD_IMAGE=arch/arm64/boot/Image
+
+generate-tests:
+  image: debian:trixie-slim
+  stage: generate
+  tags:
+    - lightweight
+  dependencies:
+    - build-arm64-debian-package
+  before_script:
+    - apt update
+    - apt install -y zstd
+  script:
+    - mkdir deb
+    - "for x in artifacts/linux-image*.deb ; do dpkg -x ${x} deb ; done"
+    - cp deb/boot/vmlinuz* vmlinuz
+    - tar -f modules.tar.zst -C deb -c --zstd -v lib/modules
+    - mkdir dtbs
+    - cp -r deb/usr/lib/linux-image*/* dtbs
+    - sed -i s,%%KERNEL_BUILD_JOB%%,${CI_JOB_ID},g lava/testjob.yaml
+  artifacts:
+    paths:
+      - vmlinuz
+      - modules.tar.zst
+      - dtbs
+      - lava/testjob.yaml
+
+.lava-test:
+  stage: lava
+  tags:
+    - lava-runner
+  rules:
+    - if: $LAVA_TOKEN
+  script:
+    - submit lava/testjob.yaml
+  dependencies:
+    - generate-tests
+  artifacts:
+    when: always
+    paths:
+      - "*"
+    reports:
+      junit: "*_junit.xml"
+
+lava-test-genio-350:
+  extends: .lava-test
+  variables:
+    DEVICE_TYPE: mt8365-genio-350-evk
+    DTB: mt8365-evk
+
+lava-test-genio-510:
+  extends: .lava-test
+  variables:
+    DEVICE_TYPE: mt8370-genio-510-evk
+    DTB: mt8370-genio-510-evk
+
+lava-test-genio-700:
+  extends: .lava-test
+  variables:
+    DEVICE_TYPE: mt8390-genio-700-evk
+    DTB: mt8390-genio-700-evk
+
+lava-test-genio-1200:
+  extends: .lava-test
+  variables:
+    DEVICE_TYPE: mt8395-genio-1200-evk
+    DTB: mt8395-genio-1200-evk
diff --git a/lava/testjob.yaml b/lava/testjob.yaml
new file mode 100644
index 000000000000..948eb8224f59
--- /dev/null
+++ b/lava/testjob.yaml
@@ -0,0 +1,46 @@
+device_type:  {{ job.DEVICE_TYPE }}
+
+job_name: MediaTek AIoT tests {{job.CI_JOB_ID}}
+timeouts:
+  job:
+    minutes: 15
+  action:
+   minutes: 5
+visibility: public
+
+context:
+  extra_kernel_args: rootwait
+
+actions:
+  - deploy:
+      timeout:
+        minutes: 2
+      to: tftp
+      kernel:
+        url: "{{job.CI_PROJECT_URL}}/-/jobs/%%KERNEL_BUILD_JOB%%/artifacts/raw/vmlinuz"
+        type: image
+      modules:
+        url: "{{job.CI_PROJECT_URL}}/-/jobs/%%KERNEL_BUILD_JOB%%/artifacts/raw/modules.tar.zst"
+        compression: zstd
+      dtb:
+        url: "{{job.CI_PROJECT_URL}}/-/jobs/%%KERNEL_BUILD_JOB%%/artifacts/raw/dtbs/mediatek/{{ job.DTB }}.dtb"
+      ramdisk:
+        url: "https://gitlab.collabora.com/mediatek/aiot/debian-image-recipes/-/jobs/artifacts/main/raw/out/trixie-arm64-minimal-initramfs.cpio.zst?job=build-lava-debian-minimal-fs"
+        compression: zstd
+      nfsrootfs:
+        url: "https://gitlab.collabora.com/mediatek/aiot/debian-image-recipes/-/jobs/artifacts/main/raw/out/trixie-arm64-minimal-rootfs.tar.zst?job=build-lava-debian-minimal-fs"
+        compression: zstd
+
+  - boot:
+      method: u-boot
+      commands: nfs
+      timeout:
+        minutes: 10
+      auto_login:
+        login_prompt: 'login:'
+        username: user
+        password_prompt: 'Password:'
+        password: user
+      prompts:
+        - 'user@debian-trixie(.*)$'
+        - 'root@debian-trixie(.*)#'
-- 
2.43.0

